### PR TITLE
fix c.spec() == nil cause nil point panic

### PR DESCRIPTION
**Pull request by BSWANG**
fix c.spec() == nil cause nil point panic on `docker run --net` at attachable network

Signed-off-by: bingshen.wbs <bingshen.wbs@alibaba-inc.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
When can it be nil?

**Comment by BSWANG** [comment]
Run containers on attachable  network

**Comment by TomSweeneyRedHat** [comment]
It would be nice to get the failing checks to pass.  Is this also potentially an issue for the mounts function near line 251 and the hostConfig function near line 365?

**Comment by BSWANG** [comment]
@TomSweeneyRedHat Thanks for your notification, I will fix those failing checks and issues.

**Comment by tophj-ibm** [comment]
@BSWANG power failing test isn't related, and I'm 99% sure the windows isn't either. I've restarted both of those jobs just in case.

**Comment by denverdino** [comment]
I hit the problem also, it will be nice for the nil check

**Review by AkihiroSuda — APPROVED** [review]
LGTM

**Comment by BSWANG** [comment]
@TomSweeneyRedHat  I add more judgement on more `c.spec()` call.

File: daemon/cluster/executor/container/container.go [hunk 1]
```diff
@@ -631,6 +647,9 @@ func (c *containerConfig) networkCreateRequest(name string) (clustertypes.Networ
 }
 
 func (c *containerConfig) applyPrivileges(hc *enginecontainer.HostConfig) {
+	if c.spec() == nil {
+		return
+	}
```
Comment (aaronlehmann) [inline daemon/cluster/executor/container/container.go]: The call to this function is already guarded by a `c.spec() != nil` check, so I don't think this is necessary.

**Comment by aaronlehmann** [comment]
What code path causes the nil pointer dereference? I'm not seeing where these methods would be called for a conventional container attached to a swarm network. `networkAttacherController` seems to only call a few methods on `containerAdapter`:

- `createNetworks`
- `networkAttach`
- `waitForDetach`
- `removeNetworks`

I don't see any of these calling the `containerConfig` methods modified in this PR.

**Comment by AkihiroSuda** [comment]
ping @BSWANG

File: daemon/cluster/executor/container/container.go (hunk 1 above)
Comment (BSWANG) [inline daemon/cluster/executor/container/container.go]: Thanks, I will remove this check.

**Comment by aaronlehmann** [comment]
I think that calling, for example, `image()` on something that doesn't have a `ContainerSpec` would be a bug. I don't really think it makes sense to add these checks everywhere in anticipation of introducing buggy code like this. With these checks, the failures would be silent, instead of causing a panic that tells us something is wrong.

The ideal solution would be to redefine `containerConfig` as an interface, and have another interface which is specific to container tasks (not network attachments) that builds on top of it. Then the compiler would make sure that only certain methods can be called in contexts where we don't have a `ContainerSpec`.

**Comment by AkihiroSuda** [comment]
ping @BSWANG

**Comment by BSWANG** [comment]
@AkihiroSuda Thanks, I think I should solve it by @aaronlehmann's suggestion. I will do it at this weekend.

**Comment by AkihiroSuda** [comment]
@BSWANG
any update?

**Comment by BSWANG** [comment]
@AkihiroSuda Sorry, I have no time to do this recently. I will close this PR. If I have any update future. I will reopen it. Thanks.

**Pull request closed without merging**